IUC want profile in the <tool> and detect_errors in the <command>

fubar2 opened the pull request:
Hi @hexylena! 
Request from the IUC to add boilerplate doodads to ToolFactory outputs, so this PR elevates command to an XMLParam, adding the command string as a CDATA text. It passes tests including the new one and more importantly, the existing one, though I'm not sure if this is the most efficient path. Whatever, it works for me FWIW - thanks!

fubar2 committed:
IUC want profile in the <tool> and detect_errors in the <command>
Only way I could get it to work was to make command an XMLParam and
add the command string as an etree.CDATA.
It passes tests including a new one for the command but I'm not sure if this is the most
efficient way to achieve this goal.

fubar2 committed:
tox ini does the needful and runs the unit tests
why not?

fubar2 committed:
cleanup. is good. is xml.

bernt-matthias reviewed (commented):
Good idea to allow setting a tool's `profile` and a command's `detect_errors` property. 
Is setting `detect_errors`  to anything else than `"aggressive"` supported?
Can you add a test showing how to set non-default values?

hexylena commented:
>  Is setting detect_errors to anything else than "aggressive" supported?

[yes, lots of options](https://docs.galaxyproject.org/en/latest/dev/schema.html#error-detection), and aggressive is generally the wrong choice, IMO. I would much, much rather see it set to `exit_code`.

Aggressive is rather annoying when wrapping other people's tools; as soon as some library decides to log a non-fatal warning or error, the job fails, despite the job in fact succeeding by exit code/results. Too many tools have used "Error:" in their outputs that I avoid aggressive.

hexylena commented:
@bernt-matthias where are you with #37, should we merge it? I'd like the github workflow that you've done there.

bernt-matthias commented:
> @bernt-matthias where are you with #37, should we merge it? I'd like the github workflow that you've done there.

Ready for review... Not sure if tests succeed yet. Can also separate the workflow in a separate PR if needed.

fubar2 commented:
Yes, defaults are both what @bgruening asked for.  
Profile and detect_errors are set by the caller like this example for profile:
```
        self.newtool = gxt.Tool(
            self.tool_name,
            self.tool_id,
            self.tool_version,
            self.args.tool_desc,
            self.args.sysexe,
            profile = self.profile
```
 and here:`
self.newtool.command = gxtp.Command(detect_errors="aggressive")`

These changes already have one new test and import_xml_sample has these elements - and they are correctly parsed. 
The ToolFactory tests them too if I use this updated galaxyxml and it passes its own tests. 
In summary: I am not sure exactly how to satisfy this beyond the already added unit test - hints appreciated. Are you suggesting another independent test @bernt-matthias? I cannot debate the defaults - I was following instructions but will leave that to others. _Please let me know explicitly what I need to do to get this in so I can get on using it?_

fubar2 committed:
Tweaks to ensure the detect_errors value is indeed controlled.
That old problem of the galaxy utility libs being stuck at 23.05 is biting.
I wonder when they will be updated and all the downstream packages?

`  File "/home/ross/rossgit/galaxyxml/.venv/lib/python3.10/site-packages/galaxy_tool_util-23.0.5-py3.10.egg/galaxy/tool_util/deps/conda_util.py", line 90, in CondaContext
    _conda_version: Optional[Union[packaging.version.Version, packaging.version.LegacyVersion]]
AttributeError: module 'packaging.version' has no attribute 'LegacyVersion'
`

fubar2 commented:
Would this additional passing unit test be sufficient - just sets and tests the new values.
`
class TestChangeProfDet(TestImport):
    def test_changes(self):
        self.tool.command.node.attrib["detect_errors"] = "foobarfoo"
        de = self.tool.command.node.attrib["detect_errors"]
        self.assertEqual(de, "foobarfoo")
        self.tool.profile  = "anything you want"
        self.assertEqual(self.tool.profile,  "anything you want")
`

fubar2 committed:
add another unit test to demonstrate mutability for profile and detect error user set values

bernt-matthias commented:
I guess the most undebatable option is to use Galaxy's defaults, ie to not set them by default.

Btw would this be done by setting the value to `None`? 

But I'm fine with any default.

fubar2 commented:
So all good @bernt-matthias ? @bgruening suggested those defaults and since I no longer see real use, I have no information on their effects - so changing them to anything else is fine with me. Generated tools will have those values once the PR is in.

bernt-matthias commented:
For your PRs I would suggest to go with marius' suggestion https://github.com/bgruening/galaxytools/pull/1326#discussion_r1314559091, ie set `profile` and no `detect_errors`

fubar2 commented:
ok. one moment please....

bernt-matthias commented:
I don't see in your coffee the possibility to omit the two attributes from the generated xml

fubar2 committed:
changed default to just profile. No detect_errors.
seems just fine.

fubar2 commented:
ok Thanks again @bernt-matthias.
Default is None for detect_errors now.
import_xml.xml now has profile and no detect_errors.
It parses with testimport.py as expected giving 22.05 profile
and an empty command tag now.
tests all pass with tox.


```
ross@pn50:~/rossgit/galaxyxml/test$ python testimport.py import_xml.xml
reading from import_xml.xml             
INFO:galaxyxml.tool.import_xml:<description> is loaded during initiation of the object.
INFO:galaxyxml.tool.import_xml:<stdio> loaded.                     
INFO:galaxyxml.tool.import_xml:<version_command> is loaded during initiation of the object.
<tool name="Import" id="import_test" version="1.0" profile="22.05">
  <description>description</description>                                                  
  <edam_operations>                      
    <edam_operation>operation_0004</edam_operation>                     
  </edam_operations>
  <edam_topics>
    <edam_topic>topic_0003</edam_topic>
  </edam_topics>                                                                          
  <requirements>     
    <requirement version="1" type="package">magic_package</requirement>
    <container type="docker">a_container</container>
  </requirements>
  <stdio>                                                                                 
    <exit_code range="1:" level="fatal"/>
  </stdio> 
  <version_command><![CDATA[v_command]]></version_command> 
  <command><![CDATA[                                                                                                                                                                
command interval_file $interval_file
$bool       
int_size $int_size
float_size $float_size
```

fubar2 commented:
>I don't see in your coffee the possibility to omit the two attributes from the generated xml
Ah. If you remove it from the sample, it parses just fine. However, the emitted xml has it as the default. Is there ever a good reason to have no profile? You can set it to anything but if it has a default, that will appear in all generated output, satisfying
https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html?highlight=profile#tool-profile

Are there situations where it will be bad in generated XML so also needing default of none?

fubar2 commented:
@hexylena @bernt-matthias - is there still a problem with this PR being merged? Please let me know what needs changing or whether you would prefer me to close it?

bernt-matthias commented on galaxyxml/tool/import_xml.py:
```diff
@@ -25,12 +25,10 @@ def _init_tool(self, xml_root):
                 description = child.text
-            elif child.tag == "command":
-                executable = child.text.split()[0]
-                command = child.text
```
Why is command removed here?

hexylena replied:
it took me a while to understand this, it's because now the command is parsed in `Command`, rather than expecting the entire thing to be set from the start. I was also unhappy it was re-ordered since it made the diff unnecessarily confusing.

fubar2 replied:
not a deliberate or useful reorder - now less reordered.

bernt-matthias commented on test/unit_test_import_xml.py:
```diff
@@ -27,15 +26,34 @@ def test_override(self):
+        try:
+            de = self.tool.command.node.attrib["detect_errors"]
+        except KeyError:
+                de = None
```
Fix indentation here.

bernt-matthias commented on test/unit_test_import_xml.py:
```diff
@@ -27,15 +26,34 @@ def test_override(self):
+    def test_changes(self):
+        self.tool.command.node.attrib["detect_errors"] = "foobarfoo"
+        de = self.tool.command.node.attrib["detect_errors"]
+        self.assertEqual(de, "foobarfoo")
```
I don't understand this test. Looks like:

```
A = "foobarfoo"
B = A
self.assertEqual(B, "foobarfoo")
```

Same for the following two lines.

fubar2 replied:
yes. Makes sure they are persistent and mutable after the previous test checks the correct values have been extracted from the sample. Was trying to respond to a previous suggestion. Will remove it.

bernt-matthias commented on test/import_xml.xml:
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0"?>
-<tool id="import_test" name="Import" version="1.0">
```
The xml examplifies the defaults, or?

hexylena commented:
> However, the emitted xml has what was supplied or the default profile if there was not one, satisfying https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html?highlight=profile#tool-profile and the rule that one should be forgiving of what is ingested but strict in what is emitted.

This library isn't intended to generate IUC best practice tools, that's not it's goal, it's intended to let you flexibly generate the tool you need. 

> Are there situations where that guideline should not be followed in generating XML? If we are not following those guidelines, it's going to be hard to make a good decision.

this is a prescriptivist take (these are our rules and they are the only right way) and this library is targetted at people that know what they want to do and want flexibility, not enforcing IUC best practices (as you'll note, it's not under galaxy-iuc)

hexylena commented on galaxyxml/tool/__init__.py:
```diff
@@ -134,49 +131,52 @@ def export(self, keep_old_command=False):
 
         if self.command_override:
-            command_line = self.command_override
+            command_text = self.command_override
```
why did you rename this?

fubar2 replied:
Plan to revert all this - part of the command attributes change.
Took me a while to get my head around where the actual command text was being found in the new XMLParam and I discovered I was reusing a variable name. Went the wrong way - sorry - should have backed out but forged ahead with new names. Will soon be back to what it was.

fubar2 replied:
The stdio default insertion is dropped. 
Stdios still work fine if you want to use them. :) 
This all started when I was asked to add profile and remove that deprecated default stdio value, in a review of some tools generated using galaxyxml, so not sure what else I can do here.

hexylena commented on galaxyxml/tool/import_xml.py:
```diff
@@ -77,21 +74,26 @@ def _load_stdio(self, tool, stdio_root):
-        :type tool: :class:`galaxyxml.tool.Tool`
-        :param desc_root: root of <command> tag.
-        :type desc_root: :class:`xml.etree._Element`
+        now an XMLParameter with a text
```
please don't delete documentation, update it if it needs updating.

hexylena commented on galaxyxml/__init__.py:
```diff
@@ -1,7 +1,4 @@
-    object,
-    str
-)
+from builtins import object, str
```
yes we should have been running `make reformat` more regularly, but in the future please don't include reformatting changes mixed in with your PR, it makes it much more difficult to review. Reformatting should be a separate PR.

fubar2 replied:
Not intentional just can't figure out how to use tox test without being reformatted :( sorry.

fubar2 replied:
ok - sorry! Unintended changes sjhould now be minimal, but tox...

hexylena commented on galaxyxml/tool/__init__.py:
```diff
@@ -134,49 +131,52 @@ def export(self, keep_old_command=False):
+                command_text.append(export_xml.outputs.cli())
             except Exception:
                 pass
-        # Steal interpreter from kwargs
-        command_kwargs = {}
-        if export_xml.interpreter is not None:
-            command_kwargs["interpreter"] = export_xml.interpreter
```
yes, ok, this should not have been included. good. https://docs.galaxyproject.org/en/latest/dev/schema.html#id22 @bernt-matthias the documentation looks a bit odd there.

hexylena commented on galaxyxml/tool/__init__.py:
```diff
@@ -134,49 +131,52 @@ def export(self, keep_old_command=False):
-            if getattr(self, "command", None):
-                command_node.text = etree.CDATA(export_xml.command)
+            if getattr(self, "command_line", None):
+                ctext = export_xml.command_text
```
now that you've renamed command_line, this should be command_text, right?

```suggestion
            if getattr(self, "command_text", None):
                ctext = export_xml.command_text
```

bernt-matthias reviewed (commented):
I think there are now some diffs in the example:

```
        python examples/example.py > tmp.xml
        xmldiff tmp.xml examples/tool.xml
        python examples/example_macros.py > tmp.xml
        xmldiff tmp.xml examples/example_macros.xml
```

Can you fix those?

hexylena commented on galaxyxml/tool/__init__.py:
```diff
@@ -134,49 +131,52 @@ def export(self, keep_old_command=False):
+                )
+            ctext = actual_cli.strip()
+        export_xml.command_text = ctext
+        ctext = '\n' + ctext # pretty - bjoern's suggestion
```
I don't love this tbh, it'll introduce lots of diffs to add some useless whitespace. If you want a newline in your command add it, if you want it pretty, run it through planemo. This isn't the job of this library

hexylena committed:
Not the job of this library

hexylena committed:
Ensure tests pass

hexylena commented on tox.ini:
```diff
@@ -1,3 +1,15 @@
 [flake8]
 max-line-length = 230
 max-complexity = 10
+[testenv]
+description = run the tests with pytest
+package = wheel
+wheel_build_env = .pkg
+deps =
+    pytest>=6
+commands =
+    pytest test/unit_test_import_xml.py {tty:--color=yes} {posargs}
```
We have not been using tox to test, this is a bigger change, hmm. until now it was all `nose`, which apparently is [now broken](https://github.com/nose-devs/nose/issues/1099). I _really_ wish some of these changes had come in separate PRs so we could've discussed them in isolation, but I guess we have to accept this because nose is non functional.

hexylena committed:
cleanup tox, pass tests

hexylena commented:
@bernt-matthias you have most of the same review comments I did :sweat_smile: I should've hit submit earlier

hexylena reviewed (commented):
As of now it doesn't pass the tests (the other tests, in the .travis.yml that we didn't finish turning into a GHA.). I will update those @fubar2 .

```diff
$ diff tmp.xml examples/tool.xml 
1c1
< <tool name="aragorn" id="se.lu.mbioekol.mbio-serv2.aragorn" version="1.2.36" profile="22.05">
---
> <tool name="aragorn" id="se.lu.mbioekol.mbio-serv2.aragorn" version="1.2.36">
7a8,11
>   <stdio>
>     <exit_code range="1:" level="fatal"/>
>   </stdio>
>   <expand macro="stdio"/>
9,10c13
<   <command><![CDATA[
< aragorn.exe $flag
---
>   <command><![CDATA[aragorn.exe $flag
```

and yeah, fine with dropping the stdio, but still not crazy about adding profile by default, that really seems like something for library consumers to do. I also don't understand the newline being added there.

hexylena committed:
cleanup the tox

hexylena commented:
I would recommand also that don't make PRs from your `master` branch in the future, it's best to make them from a feature branch.

hexylena commented:
@fubar2 @bernt-matthias 

```
        python examples/example.py > tmp.xml
        xmldiff tmp.xml examples/tool.xml
        python examples/example_macros.py > tmp.xml
        xmldiff tmp.xml examples/example_macros.xml
```

I fixed the first set, I'm still struggling to figure out why the second set changed, @fubar2 maybe you can fix that.

fubar2 commented:
ok - thanks for the comments - appreciated.
 
Most problems are from changes to make <command>  an XMLParam. 
Please let me know if you really want to leave them in this PR - would prefer to revert them otherwise.

Apologies for the import sample changes - from using it to confirm that either missing or supplied values were handled - shouldn't have committed those.
Yes, too much at once and yes, a branch next time after this reminder...
I'll try to figure the rest out and make adjustments in response after I revert the command stuff?

hexylena commented:
It happens, no worries. Let's not dwell.

- Please feel free to update this PR accordingly. 
- Making it a proper `Command` is a good change and one that I would like to see kept, as I see it that's one of the core improvements in this PR.
- If the tests pass, well, that's what's really important in the end.

fubar2 commented:
ok. 
Thanks again!

fubar2 committed:
trying to undo lots of overdo.

fubar2 committed:
not saved..

fubar2 committed:
reformat during tests...

fubar2 committed:
get rid of notional test..

fubar2 committed:
revert order in init

fubar2 commented:
> > However, the emitted xml has what was supplied or the default profile if there was not one, satisfying https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html?highlight=profile#tool-profile and the rule that one should be forgiving of what is ingested but strict in what is emitted.
> 
> This library isn't intended to generate IUC best practice tools, that's not it's goal, it's intended to let you flexibly generate the tool you need.
> 
> > Are there situations where that guideline should not be followed in generating XML? If we are not following those guidelines, it's going to be hard to make a good decision.
> 
> this is a prescriptivist take (these are our rules and they are the only right way) and this library is targetted at people that know what they want to do and want flexibility, not enforcing IUC best practices (as you'll note, it's not under galaxy-iuc)

Apologies - my immediate problem is to get things past review elsewhere - sorry it spilled over where it does not belong.  I'll  advocate for the changed defaults recommended by those reviewers, but all those decisions are entirely yours...

fubar2 committed:
eeesh. Found a bad refactor booboo.

fubar2 commented:
I hope all the worst is gone and it's ready for another review. 
Reformatting is from running tests in the dumbest possible way so not intended but effective.
Sorry for the noise. 
Defaults for profile and command detect_errors are None.

hexylena commented:
>  The IUC guidance seems way out of date and hard to know which bits to take seriously. Disappointing and hard to support really. A condign punishment.

It would be great to get it updated but, things move very very slowly there it seems. c.f. my indentation PR allowing tabs, which is now dormant since months. We also didn't have the 'traditional' IUC meeting at GCC too, so, didn't have the usual going through of issues/etc.

hexylena reviewed (approved):
This round was much better to review, thanks for changes @fubar2.

fubar2 committed:
renamed unit test back to unit_test.. but that stops it testing - needs to start with test for tox to run automagically it seems.
Reverted the sample

fubar2 committed:
reformat

fubar2 committed:
mostly my own flake things

fubar2 commented:
github action flake now unhappy so format issues to resolve or silence?

hexylena committed:
remove stdio from ref

hexylena commented:
Feel free to ignore flake. I've pushed a commit to take care of one test, then there's only one more failure:

```
12c12,13
<   <token name="ARAGORN_OUTMACRO"><![CDATA[]]></token>
---
>   <token name="ARAGORN_OUTMACRO"><![CDATA[-output '$output'
> ## TODO CLI for OutputCollection collection]]></token>
```

Previously the outmacro had included an `-output '$output'` with a TODO for the collection, now that seems to be missing, but I think this was an issue already for a while. You contributed the collection support ages ago (awesome thank you :yellow_heart:) but I think our tests haven't been working since then so we didn't notice until now.

So I will merge over this specific failure, and maybe if you have energy and bandwidth, if you have a nice fix for it, i think it would be useful to the library's consumers to retain this todo/-output

<details><summary>flake rant</summary>
 It's stuff black will take care of which is especially, especially frustrating for me. Why lint and make a human fix it, just fix it and shut up (I say to flake/black.) These ridiculous tools should either fix it automatically because it's bloody whitespace, or stop complaining. `isort` is banned from all of my projects because it drives me up a wall.
</details>

hexylena committed:
whatever

hexylena commented:
Thanks for the patience @fubar2 and putting up with our complaints. I'll mint a new release for ya

hexylena commented:
after a silly amoutn of fighting with pypi/gha, https://pypi.org/project/galaxyxml/0.5.3/ now exists.
